### PR TITLE
c-based: fix to handle edge case

### PR DESCRIPTION
**Pull request by jafl**
where "character" returned is a symbol (> 0xff)

### CODE REVIEW

**Comment by masatake** [comment]
Thank you.
Related to #3771.  I'm working on this topic very slowly.

**Comment by jafl** [comment]
@masatake I can tackle this issue and try to finish it.  Clearly, anywhere that calls `cppGetc()` needs to be updated to handle c > 0xff.  What other pieces are there that need to be fixed?

**Comment by jafl** [comment]
I simplified my original changes, because I realized that `cppUngetc` is just fine with c > 0xff

**Comment by jafl** [comment]
Failed checks are system issues, not code.

**Comment by leleliu008** [comment]
Github Actions may fail for network issues, If you encounter this situation next time, just rerun it. It usually would become green after rerunning.

**Comment by jafl** [comment]
@masatake Here is my proposal for fixing the issue.  I created `cStringPut` as a smart wrapper around `vStringPut` and made `cppUngetc` smarter.  If you are ok with this, I will convert all `vStringPut` to `cStringPut` in `./cxx/cxx_parser_tokenizer.c`, `cpreprocessor.c`, `vera.c`, and `c-based.c`.

I only changed the minimal number of calls so far, to minimize the size of the commit and make it easier to review.

**Comment by masatake** [comment]
Give me some time.

**Comment by masatake** [comment]
@jafl Looks fine.
Before merging this, I would like to merge #3804.
Could you consider updating cStringPut in your pull request after I merge #3804?

**Comment by jafl** [comment]
@masatake Sure.  Let me know when you merge #3804, and I'll rebase.

**Comment by jafl** [comment]
@masatake Is it ok if I convert all `vStringPut` to `cStringPut` in all the files listed in #3771 ?

File: parsers/cpreprocessor.h [hunk 1]
```diff
@@ -99,6 +99,7 @@ extern int cppUngetBufferSize();
 extern void cppUngetString(const char * string,int len);
 extern int cppGetc (void);
 extern const vString * cppGetLastCharOrStringContents (void);
+extern void cStringPut (vString * string, const int c);
```
Comment (masatake) [inline parsers/cpreprocessor.h:102]: Please add a comment explaining how cStringPut differs from vStringPut.

**Comment by masatake** [comment]
I merge #3804. Could you rebase this?
You must update the definition of cStringPut.

> @masatake Is it ok if I convert all vStringPut to cStringPut in all the files listed in https://github.com/universal-ctags/ctags/issues/3771 ?

At this time, could you update only vStringPut in c-based.c?
I don't have enough time to review the changes for other files at once.
I deleted c.c. Even you updated c-based.c, users get the benefit of the change.

For fixing #3771, I have been thinking about the way of handling EOF.
When I read this pull request, I recognized that thinking about EOF is insufficient. Thank you for the pull request.

I have been finding my mistakes in cpreprocess.c because of my limited understanding of the file.

**Review by masatake — APPROVED** [review]

**Comment by masatake** [comment]
Thank you.

**Pull request merged**